### PR TITLE
Clean-up: be consistent in action-reason structure to lint messages

### LINKED ISSUE #1330: Clean-up: fine-tune linter messages to be in "Action, reason" order where possible
_Originally posted by @MichaelChirico in https://github.com/r-lib/lintr/pull/1311#discussion_r884201937_

My instinct is we should prefer the actionable bit up front, with explanations in the message only for clarification in cases where we can't provide laser-focused alternatives/diagnosis. The thinking being if there are dozens of lints on a file / package, the output will be easier to read/delint if the user can snap to the beginning of each message to find their action item.

OTOH I think we're a bit inconsistent on this front overall. Here are the ~120 existing messages produced by our test suite (excluding error lints, and lightly edited to condense things a bit):

```
<<1 | 1L>>:<<dim(...)[1L] | dim(...)[[1L]] | length | ncol | nrow | NCOL | NROW>> is likely to be wrong in the empty edge case. Use <<seq_len | seq_along>>() instead.
!all(x) is better than any(!x). The former applies negation only once after aggregation instead of many times for each element of x.
anyDuplicated(DF$col) == 0L is better than length(unique(DF$col)) == nrow(DF)
anyDuplicated(x, ...) > 0 is better than any(duplicated(x), ...).
anyDuplicated(x) == 0L is better than length(unique(x)) == length(x).
Any function spanning multiple lines should use curly braces.
anyNA(x) is better than any(is.na(x)).
!any(x) is better than all(!x). The former applies negation only once after aggregation instead of many times for each element of x.
Avoid storing placeholder tests like expect_equal(1, 1)
<< <<- | ->> >> can have hard-to-predict behavior; prefer assigning to a specific environment instead (with assign() or <-).
Closing curly-braces should always be on their own line, unless they are followed by an else.
Code and comments coming after a top-level return() or stop() should be removed.
Combine inputs to vectorized functions first to take full advantage of vectorization, e.g., %s(c(x, y)) only runs the more expensive %s() once as compared to c(%s(x), %s(y)).
Commas should always have a space after.
Commas should never have a space before.
Commented code should be removed.
Compound semicolons are discouraged. Replace them by a newline.
Conditional expressions require scalar logical operators (&& and ||)
Do not place spaces after parentheses.
Do not place spaces after square brackets.
Do not place spaces before parentheses.
Do not place spaces before square brackets.
Do not use absolute paths.
Don't alter the search() path in <<.onAttach | .onLoad>>() by calling <<library | require>>().
Don't slow down package load by running installed.packages() in .onLoad().
Don't use cat() in .onLoad().
Don't use <<F | T>> as a variable name, as it can break code relying on <<F | T>> being <<FALSE | TRUE>>.
Don't use message() in .onAttach().
Don't use nested <<ifelse | fifelse | if_else.>() calls; instead, try (1) data.table::fcase; (2) dplyr::case_when; or (3) using a lookup table.
Don't use <<paste | paste0>> to build <<stop | warning | message | packageStartupMessage>> strings. Instead use the fact that these functions build condition message strings from their input (using "" as a separator). For translateable strings, prefer using gettextf().
Don't use print() in .onLoad().
Don't use writeLines() in .onAttach().
Duplicate arguments in function call.
Either both or neither branch in `if`/`else` should use curly braces.
`else` should come on the same line as the previous `}`.
expect_false(x) is better than expect_identical(x, FALSE)
expect_false(x) is better than expect_true(!x), and vice versa.
expect_gte(x, y) is better than expect_true(x >= y).
expect_gt(x, y) is better than expect_true(x > y).
expect_identical(x, y) is better than expect_true(x == y).
expect_length(x, n) is better than <<expect_equal | expect_identical>>(length(x), n)
expect_lte(x, y) is better than expect_true(x <= y).
expect_lt(x, y) is better than expect_true(x < y).
expect_named(x, n) is better than <<expect_equal | expect_identical>>(names(x), n)
expect_null(x) is better than <<expect_equal | expect_identical>><<(x, NULL) | (is.null(x))>>
expect_s3_class(x, k) is better than expect_equal(class(x), k). Note also expect_s4_class() available for testing S4 objects.
expect_s3_class(x, k) is better than expect_identical(class(x), k). Note also expect_s4_class() available for testing S4 objects.
expect_s3_class(x, k) is better than expect_true(is.<k>(x)) or expect_true(inherits(x, k)). Note also expect_s4_class() available for testing S4 objects.
expect_s4_class(x, k) is better than expect_true(is(x, k)). Note also expect_s3_class() available for testing S3 objects.
expect_true(x) is better than expect_equal(x, TRUE)
expect_type(x, t) is better than expect_equal(typeof(x), t)
expect_type(x, t) is better than expect_identical(typeof(x), t)
expect_type(x, t) is better than expect_true(is.<t>(x))
For static regular expression patterns, set `fixed = TRUE`. Note that this includes regular expressions that can be expressed as fixed patterns, e.g. [.] is really just . and \$ is really just $ if there are no other regular expression specials. For functions from the 'stringr' package, the way to declare a static string is to wrap the pattern in stringr::fixed(). If this is being used in a dbplyr context (i.e., translated to sql), replace the regular expression with the `LIKE` operator using the `%LIKE%` infix function. Lastly, take care to remember that the `replacement` argument of `gsub()` is affected by the `fixed` argument as well.
Function "%s" is undesirable.
Functions should have cyclomatic complexity of less than %d, this has %d.
Include the leading zero for fractional numeric constants.
Instead of comparing class(x) with <<%in% | == | !=>>, use inherits(x, 'class-name') or is.<class> or is(x, 'class')
Instead of <<assert_that | expect_false | expect_true | stopifnot>>(A || B), write multiple expectations like %s(A) and %s(B) The latter will produce better error messages in the case of failure.
Integers should not be implicit. Use the form 1L for integers or 1.0 for doubles.
Just use the logical condition (or its negation) directly instead of calling <<ifelse | fifelse | if_else>>fifelse(x, %s, %s)
.Last.lib() should take one argument starting with 'lib'.
length(...):1 is likely to be wrong in the empty edge case. Use seq_along() instead.
Lines should not be more than %d characters.
local variable '%s' assigned but may not be used
Missing argument in function call.
Missing terminal newline.
no visible <<binding for global variable | global function definition>> '%s'
.onAttach() should take two arguments, with the first starting with 'lib' and the second starting with 'pkg'.
.onDetach() should take one argument starting with 'lib'.
.onLoad() should take two arguments, with the first starting with 'lib' and the second starting with 'pkg'.
Only use double-quotes.
Opening curly braces should <<always be followed by a new line... | never go on their own line...>>.
Operator `%s` is undesirable. << As an alternative... | >>
Package '%s' is <<attached but never used | only used by namespace...>>.
paste0(...) is better than paste(..., sep = "").
Place a space before left parenthesis, except in a function call.
<<pmin | pmax>>(x, y) is preferable to <<ifelse | fifelse | if_else>(x %s y, %s, %s).
Prefer as.numeric(x) to <<ifelse | fifelse | if_else>>(x, %d, %d) if really needed, but do note that R will usually convert logical vectors to 0/1 on the fly when needed.
Prefer grep(pattern, x, ..., value = TRUE) over x[grep(pattern, x, ...)] and x[grepl(pattern, x, ...)].
Prefer stringr::str_subset(x, pattern) over x[str_detect(x, pattern)] and x[str_which(x, pattern)].
Put exactly one space on each side of infix operators.
Put library.dynam() calls in .onLoad, not .onAttach().
Put packageStartupMessage() calls in .onAttach(), not .onLoad().
Put spaces around all infix operators.
Remove spaces before the left parenthesis in a function call.
'%s' <<does not exist | is not exported from>> in {%s}.
'%s' is exported from {%s}. Use %s::%s instead.
sep= is not a formal argument to paste0(); did you mean to use paste(), or collapse=?
`%>%` should always have a space before it and a new line after it, unless the full pipeline fits on one line.
Tests should compare objects in the order 'actual', 'expected', not the reverse. For example, do %s(foo(x), 2L) instead of %s(2L, foo(x)).
There should be a space before an opening curly brace.
There should be a space between right parenthesis and a body expression.
There should be a space between right parenthesis and an opening curly brace.
This code relies on the default value of stringsAsFactors, which changed in version R 4.0. Please supply an explicit value for stringsAsFactors for this code to work with versions of R both before and after this switch.
TODO comments should be removed.
toString(.) is more expressive than paste(., collapse = ", "). Note also glue::glue_collapse() and and::and() for constructing human-readable / translation-friendly lists
Trailing blank lines are superfluous.
Trailing semicolons are not needed.
Trailing whitespace is superfluous.
%s (R %s) is not available for dependency R >= %s.
Unify consecutive calls to stopifnot().
Unneeded concatenation of a constant. Remove the "c" call.
Unneeded concatenation without arguments. Replace the "c" call by NULL or, whenever possible, vector() seeded with the correct type and/or length.
Use expect_identical(x, y) by default; resort to expect_equal() only when needed, e.g. when setting ignore_attr= or tolerance=.
Use explicit calls in magrittr pipes, i.e., `a %>% foo` should be `a %>% foo()`.
Use file.path() to construct portable file paths.
Use `[[` instead of `%s` to extract an element.
Use is.na for comparisons to NA (not == or !=)
Use library.dynam.unload() calls in .onUnload(), not .Last.lib().
Use library.dynam.unload() calls in .onUnload(), not .onDetach().
Use literals directly where possible, instead of coercion. c.f. 1L instead of as.integer(1), or NA_real_ instead of as.numeric(NA).
Use <-, not %s, for assignment.
Use spaces to indent, not tabs.
Use <<startsWith() | endsWith()>> to detect <<a fixed initial | an initial | a fixed terminal | a terminal>> substring.
Use the `...` argument of system.file() to expand paths, e.g. system.file("data", "model.csv", package = "myrf") instead of <<file.path(system.file(...)) | system.file(file.path(...))>>
Use <<TRUE | FALSE>> instead of the symbol <<T | F>>.
Variable and function names should not be longer than %d characters.
Variable and function name style should be %s.
```

I think "action, reason" is more common than "reason, action".

WDYT?

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #1330.

### CODE REVIEW

**Review by IndrajeetPatil — CHANGES_REQUESTED** [review]
Overall, I really like the active voice framing of the new messages!

File: R/assignment_linter.R [hunk 1]
```diff
@@ -111,7 +111,7 @@ assignment_linter <- function(allow_cascading_assign = TRUE,
     operator <- xml_text(bad_expr)
     lint_message_fmt <- rep("Use <-, not %s, for assignment.", length(operator))
     lint_message_fmt[operator %in% c("<<-", "->>")] <-
-      "%s can have hard-to-predict behavior; prefer assigning to a specific environment instead (with assign() or <-)."
+      "Avoid %s by assigning to a specific environment (with assign() or <-) to avoid hard-to-predict behavior."
```
[the remark this answers is not in the document]
Comment (AshesITR) [reply R/assignment_linter.R:114]: Maybe `Replace <<- by assigning to a specific environment ...`?

File: R/if_not_else_linter.R [hunk 2]
```diff
@@ -91,10 +91,7 @@ if_not_else_linter <- function(exceptions = c("is.null", "is.na", "missing")) {
     if_lints <- xml_nodes_to_lints(
       if_expr,
       source_expression = source_expression,
-      lint_message = paste(
-        "In a simple if/else statement,",
-        "prefer `if (A) x else y` to the less-readable `if (!A) y else x`."
-      ),
+      lint_message = "Prefer `if (A) x else y` to the less-readable `if (!A) y else x` in a simple if/else statement.",
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/if_not_else_linter.R:94]: Do we have an issue around that already? I thought so but didn't find it recently. If not, would you mind authoring the issue? I think you are the bigger advocate :)

[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/if_not_else_linter.R:94]: BTW, now I tend to use \` only when it removes ambiguity of mixing code with text, e.g. this is usually not needed for a single function call, but useful here since the code uses a lot of whitespace and has common English words `if`/`else`; the `is.numeric(.) || is.integer(.)` case is more borderline.

Agree that styling this with {cli} will allow us to be more consistent without considering tradeoffs like this.

File: NEWS.md [hunk 3]
```diff
@@ -11,6 +11,7 @@
    + Linters `closed_curly_linter()`, `open_curly_linter()`, `paren_brace_linter()`, and `semicolon_terminator_linter()`.
    + Helper `with_defaults()`.
 * `all_linters()` has signature `all_linters(..., packages)` rather than `all_linters(packages, ...)` (#2332, @MichaelChirico). This forces `packages=` to be supplied by name and will break users who rely on supplying `packages=` positionally, of which we found none searching GitHub.
+* Adjusted various linters' messages for consistency in readability (#1330, @MichaelChirico). In general, we favor lint messages to be phrased like "Action, reason" to but the "what" piece of the message front-and-center. This may be a breaking change for code that tests the specific phrasing of lints.
```
[the remark this answers is not in the document]
Comment (IndrajeetPatil) [reply NEWS.md:14]: I am wondering if we should have some developer-focused docs on what's the best way to integrate `{lintr}` in their development workflow?

[the remark this answers is not in the document]
Comment (AshesITR) [reply NEWS.md:14]: How does one suggest a commit without making it?

I'd replace `linters' messages` by `lint messages` for easier reading.

[the remark this answers is not in the document]
Comment (IndrajeetPatil) [reply NEWS.md:14]: > How does one suggest a commit without making it?

````
```suggestion
```
````

File: NEWS.md [hunk 4]
```diff
@@ -11,11 +11,12 @@
...
```
Comment (AshesITR) [inline NEWS.md:14]:
```suggestion
* Adjusted various lint messages for consistency in readability (#1330, @MichaelChirico). In general, we favor lint messages to be phrased like "Action, reason" to but the "what" piece of the message front-and-center. This may be a breaking change for code that tests the specific phrasing of lints.
```

File: NEWS.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply NEWS.md:14]: there's also this UI button:

image

[the remark this answers is not in the document]
Comment (MichaelChirico) [reply NEWS.md:14]: @IndrajeetPatil mine was a typo for "to put", which do you prefer?

File: R/commas_linter.R [hunk 5]
```diff
@@ -84,15 +84,15 @@ commas_linter <- function(allow_trailing = FALSE) {
...
+      lint_message = "Commas should never follow a space.",
       range_start_xpath = "number(./preceding-sibling::*[1]/@col2 + 1)", # start after preceding expression
       range_end_xpath = "number(./@col1 - 1)" # end before comma
     )
 
     after_lints <- xml_nodes_to_lints(
       xml_find_all(xml, xpath_after),
       source_expression = source_expression,
-      lint_message = "Commas should always have a space after.",
+      lint_message = "Commas should always precede a space.",
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/commas_linter.R:95]: Hmm it parses a bit odd to me to have 'space' singular but 'commas' plural. But also "Put spaces after commas" may be ambiguous (we really mean "one space"). "Put a space after a comma"? "Put one space after a comma"?

File: R/conjunct_test_linter.R [hunk 6]
```diff
@@ -127,13 +127,13 @@ conjunct_test_linter <- function(allow_named_stopifnot = TRUE,
...
-      "write multiple conditions like %s(A, B)."
+      "Write multiple expectations like %1$s(A) and %1$s(B)",
+      "Write multiple conditions like %s(A, B)"
     )
     lint_message <- paste(
-      sprintf("Instead of %s(A %s B),", matched_fun, operator),
       # as.character() needed for 0-lint case where ifelse(logical(0)) returns logical(0)
       sprintf(as.character(replacement_fmt), matched_fun),
+      sprintf("instead of %s(A %s B).", matched_fun, operator),
       "The latter will produce better error messages in the case of failure."
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/conjunct_test_linter.R:137]: > The latter will produce better error messages in the case of failure.

You're referring to this fragment? I would keep it, it's the "reason" for this lint.

File: R/duplicate_argument_linter.R [hunk 7]
```diff
@@ -58,7 +58,7 @@ duplicate_argument_linter <- function(except = c("mutate", "transmute")) {
     xml_nodes_to_lints(
       unlist(all_arg_nodes, recursive = FALSE)[unlist(is_duplicated)],
       source_expression = source_expression,
-      lint_message = "Duplicate arguments in function call.",
+      lint_message = "Remove duplicate arguments in function call.",
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/duplicate_argument_linter.R:61]: "Avoid duplicate arguments in function calls"?

File: R/namespace_linter.R [hunk 8]
```diff
@@ -157,7 +157,7 @@ build_ns_get_int_lints <- function(packages, symbols, symbol_nodes, namespaces,
     symbol_nodes[exported],
     source_expression = source_expression,
     lint_message =
-      sprintf("'%1$s' is exported from {%2$s}. Use %2$s::%1$s instead.", symbols[exported], packages[exported]),
+      sprintf("Use %2$s::%1$s, not %2$s:::%1$s.", symbols[exported], packages[exported]),
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/namespace_linter.R:160]: Maybe 'Don't use `:::` to access "%s" which is exported from "%s"'?

I'm thinking suggesting `::` might be wrong too -- the package may be attached in which case even `::` is not needed.

File: R/nested_pipe_linter.R [hunk 9]
```diff
@@ -76,7 +76,7 @@ nested_pipe_linter <- function(
     xml_nodes_to_lints(
       bad_expr,
       source_expression = source_expression,
-      lint_message = "Don't nest pipes inside other calls.",
+      lint_message = "Avoid nesting pipes inside other calls.",
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/nested_pipe_linter.R:79]: Consistency, although we could do with more:

```
$ grep -Fr '"Don\'t' R
R/T_and_F_symbol_linter.R:      make_lints(bad_assignment, "Don't use %2$s as a variable name, as it can break code relying on %2$s being %1$s.")
R/for_loop_index_linter.R:  lint_message = "Don't re-use any sequence symbols as the index symbol in a for loop."
R/condition_message_linter.R:      "Don't use", inner_call, "to build", outer_call, "strings.",
R/unnecessary_placeholder_linter.R:        "Don't use the placeholder (`.`) when it's not needed,",
R/unused_import_linter.R:        "Don't attach package '", unused_packages, "', which is only used by namespace. ",
R/package_hooks_linter.R:    lint_message <- sprintf("Don't use %s() in %s().", call_name, hook)
R/package_hooks_linter.R:      sprintf("Don't slow down package load by running installed.packages() in %s().", library_require_hook)
R/package_hooks_linter.R:      sprintf("Don't alter the search() path in %s() by calling %s().", library_require_hook, library_require_bad_call)
R/unnecessary_nested_if_linter.R:    "Don't use nested `if` statements,",

$ grep -Fr '"Avoid' R
R/assignment_linter.R:      "Avoid the assignment pipe %s; prefer using <- and %%>%% separately."
R/yoda_test_linter.R:      sprintf("Avoid storing placeholder tests like %s(1, 1)", matched_call)
R/pipe_return_linter.R:    "Avoid return() as the final step of a magrittr pipeline. ",
R/nested_pipe_linter.R:      lint_message = "Avoid nesting pipes inside other calls.",
R/nested_ifelse_linter.R:      sprintf("Avoid nested %s() calls.", matched_call),
R/implicit_integer_linter.R:        "Avoid implicit integers. Make the type explicit by using the form 1L for integers or 1.0 for doubles.",
R/one_call_pipe_linter.R:      lint_message = paste0("Avoid pipe ", pipe, " for expressions with only a single call."),
R/implicit_assignment_linter.R:      "Avoid implicit assignments in function calls.",
R/undesirable_operator_linter.R:    lint_message <- sprintf("Avoid undesirable operator `%s`.", operator)
```

Or do you think `Don't` and `Avoid` can both work?

File: R/nested_ifelse_linter.R [hunk 10]
```diff
@@ -94,8 +94,8 @@ nested_ifelse_linter <- function() {
 
     matched_call <- xp_call_name(bad_expr)
     lint_message <- paste(
-      sprintf("Don't use nested %s() calls;", matched_call),
-      "instead, try (1) data.table::fcase; (2) dplyr::case_when; or (3) using a lookup table."
+      sprintf("Avoid nested %s() calls.", matched_call),
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/nested_ifelse_linter.R:97]: Same as below (striving for consistency), let's discuss there.

File: R/implicit_integer_linter.R [hunk 11]
```diff
@@ -60,7 +60,8 @@ implicit_integer_linter <- function(allow_colon = FALSE) {
     xml_nodes_to_lints(
       numbers[is_implicit_integer(xml_text(numbers))],
       source_expression = source_expression,
-      lint_message = "Integers should not be implicit. Use the form 1L for integers or 1.0 for doubles.",
+      lint_message =
+        "Avoid implicit integers. Make the type explicit by using the form 1L for integers or 1.0 for doubles.",
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/implicit_integer_linter.R:64]: There were a few places that looked like they would benefit from customization but I avoided touching that in this (already hefty) PR. Let's file follow-ups.

[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/implicit_integer_linter.R:64]: #2388

File: DESCRIPTION [hunk 12]
```diff
@@ -40,7 +40,7 @@ Suggests:
     httr (>= 1.2.1),
     jsonlite,
     mockery,
-    patrick,
+    patrick (>= 0.2.0),
```
Comment (MichaelChirico) [inline DESCRIPTION:43]: this version (>1 year old) lets us get auto-named tests instead of redundantly doing `.test_name = x, param = x` (used in `nested_ifelse_linter()` here)

File: R/duplicate_argument_linter.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/duplicate_argument_linter.R:61]: Also #2389

File: R/vector_logic_linter.R [hunk 13]
```diff
@@ -89,7 +89,7 @@ vector_logic_linter <- function() {
     xml_nodes_to_lints(
       bad_expr,
       source_expression = source_expression,
-      lint_message = "Conditional expressions require scalar logical operators (&& and ||)",
+      lint_message = "Use scalar logical operators (&& and ||) in conditional expressions.",
```
Comment (MichaelChirico) [inline R/vector_logic_linter.R:92]: This should reference the specific corresponding operator (`&`->`&&`, `|`->`||`)

Comment (MichaelChirico) [reply R/vector_logic_linter.R:92]: #2390

File: NEWS.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (IndrajeetPatil) [reply NEWS.md]: > @IndrajeetPatil mine was a typo for "to put", which do you prefer?

Ah, I see. Then you can use "to put".

[the remark this answers is not in the document]
Comment (AshesITR) [reply NEWS.md]: > there's also this UI button:
> 
> 
> 
> image
> 
> 

Sadly, not on mobile 😄

File: R/commas_linter.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (AshesITR) [reply R/commas_linter.R:95]: Put a space after a comma sounds fine to me.

File: R/conjunct_test_linter.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (AshesITR) [reply R/conjunct_test_linter.R:137]: All right.

File: R/implicit_integer_linter.R (hunk 11 above)
[the remark this answers is not in the document]
Comment (AshesITR) [reply R/implicit_integer_linter.R:64]: I'd still shorten the message here (Suggestion: "Avoid implicit integers. Use 1L for integers or 1.0 for doubles.")

File: R/namespace_linter.R (hunk 8 above)
[the remark this answers is not in the document]
Comment (AshesITR) [reply R/namespace_linter.R:160]: Also fine. Re attached: There's a separate linter for that, no?

File: R/nested_pipe_linter.R (hunk 9 above)
[the remark this answers is not in the document]
Comment (AshesITR) [reply R/nested_pipe_linter.R:79]: I think both can work. Don't feels a bit stronger, but I don't mind the variation.
Also note that in printed output, this consistency would create barrages of "Avoid" making it harder to distinguish lint types at a glance.

File: R/namespace_linter.R (hunk 8 above)
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/namespace_linter.R:160]: Not really... I think this same linter will throw on `library()` for packages that are _always_ used with `::`, but nothing if there's a mix:

```r
lint("
library(lintr)
lintr::duplicate_argument_linter()
assignment_linter()
", linters = modify_defaults(linters_with_tags(NULL), undesirable_function_linter = NULL))
```

File: R/nested_pipe_linter.R (hunk 9 above)
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply R/nested_pipe_linter.R:79]: OK, reverting to keep the diff smaller.

File: R/is_numeric_linter.R [hunk 14]
```diff
@@ -97,7 +97,7 @@ is_numeric_linter <- function() {
       class_expr,
       source_expression = source_expression,
       lint_message = paste(
-        'is.numeric(x) is the same as class(x) %in% c("integer", "numeric").',
+        'Use is.numeric(x) instead of the equivalent class(x) %in% c("integer", "numeric").',
```
Comment (AshesITR) [inline R/is_numeric_linter.R:100]: Actually, this is factually incorrect.
``` r
tmp <- structure(1.0, class = "foo")
is.numeric(tmp)
#> [1] TRUE
class(tmp) %in% c("integer", "numeric")
#> [1] FALSE
```

<sup>Created on 2023-12-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Comment (MichaelChirico) [reply R/is_numeric_linter.R:100]: Good point. Just drop "the equivalent" then? "Use is.numeric() instead of class()..."?

Comment (AshesITR) [reply R/is_numeric_linter.R:100]:
```suggestion
        'Use is.numeric(x) instead of class(x) %in% c("integer", "numeric"), or `inherits(x, "integer")` if you need to check classes.',
```

Comment (MichaelChirico) [reply R/is_numeric_linter.R:100]: I worry that's a distraction... needing `is.numeric(x) && inherits(x, c("integer", "numeric"))` should be pretty rare + advanced usage.

Comment (AshesITR) [reply R/is_numeric_linter.R:100]: That's true, and it's at the end of the lint message for this. I feel it adds value because it also teaches that you shouldn't use `class(x) %in% ...` in general.

Comment (MichaelChirico) [reply R/is_numeric_linter.R]: we've already got `class_equals_linter()` for that 😉

```r
lint("class(x) %in% c('integer', 'numeric')\n", class_equals_linter())
# <text>:1:1: warning: [class_equals_linter] Instead of comparing class(x) with %in%, use inherits(x, 'class-name') or is.<class> or is(x, 'class')
# class(x) %in% c('integer', 'numeric')
# ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (AshesITR) [reply R/is_numeric_linter.R]: Ah, thanks for pointing that out. Then LGTM 😁

**Review by AshesITR — APPROVED** [review]

**Review by AshesITR — APPROVED** [review]

**Comment by MichaelChirico** [comment]
@IndrajeetPatil you have to mark your request change as fulfilled (assuming that's done)

**Review by IndrajeetPatil — APPROVED** [review]
@MichaelChirico Sorry, I wanted to do another pass at the PR after @AshesITR's review but I am not sure if and when I will find time and I don't want to keep this on hold.

**Comment by MichaelChirico** [comment]
> @MichaelChirico Sorry, I wanted to do another pass at the PR after @AshesITR's review but I am not sure if and when I will find time and I don't want to keep this on hold.

Feel free to do so at your leisure, we can always keep iterating better messages
